src/Symfony/Component/Yaml/Tests/InlineTest.php

```diff
@@ -736,4 +736,69 @@ public function getTestsForOctalNumbers()
             'negative octal number' => [-28, '-034'],
         ];
     }
+
+    /**
+     * @dataProvider unquotedExclamationMarkThrowsProvider
+     */
+    public function testUnquotedExclamationMarkThrows($value)
+    {
+        $this->expectException(ParseException::class);
+        $this->expectExceptionMessageRegExp('/^Using the unquoted scalar value "!" is not supported\. You must quote it at line 1/');
+
+        Inline::parse($value);
+    }
+
+    public function unquotedExclamationMarkThrowsProvider()
+    {
+        return [
+            ['!'],
+            ['! '],
+            ['!  '],
+            [' ! '],
+            ['[!]'],
+            ['[! ]'],
+            ['[!  ]'],
+            ['[!, "foo"]'],
+            ['["foo", !, "ccc"]'],
+            ['{foo: !}'],
+            ['{foo:     !}'],
+            ['{foo: !, bar: "ccc"}'],
+            ['{bar: "ccc", foo: ! }'],
+            ['!]]]'],
+            ['!}'],
+            ['!,}foo,]'],
+            ['! [!]'],
+        ];
+    }
+
+    /**
+     * @dataProvider quotedExclamationMarkProvider
+     */
+    public function testQuotedExclamationMark($expected, $value)
+    {
+        $this->assertSame($expected, Inline::parse($value));
+    }
+
+    // This provider should ideally stay consistent with unquotedExclamationMarkThrowsProvider
```

Comment:
let's remove "ideally" - it conveys only doubts as far as maintainance is concerned